tests/optimizers/test_general_optimizer.py

```diff
@@ -0,0 +1,242 @@
+    with pytest.raises(KeyError):
+        GeneralOptimizerPSO(5, 2, options, Random(static=static))
+
+
+@pytest.mark.parametrize(
+    "options",
+    [
+        {"c2": 0.7, "w": 0.5, "k": 2},
+        {"c1": 0.5, "w": 0.5, "k": 2},
+        {"c1": 0.5, "c2": 0.7, "k": 2},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5},
+    ],
+)
+def test_keyword_exception_random(options):
+    """Tests if exceptions are thrown when keywords are missing and a Random topology is chosen"""
+    with pytest.raises(KeyError):
+        GeneralOptimizerPSO(5, 2, options, Random())
+
+
+@pytest.mark.parametrize(
+    "options",
+    [
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": -1, "p": 2},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 6, "p": 2},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 2, "p": 5},
+    ],
+)
+def test_invalid_k_or_p_values(options, static):
+    """Tests if exception is thrown when passing
+    an invalid value for k or p when using a Ring topology"""
+    with pytest.raises(ValueError):
+        GeneralOptimizerPSO(5, 2, options, Ring(static=static))
+
+
+@pytest.mark.parametrize(
+    "options",
+    [
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": -1},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 6},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 0.5}
+    ],
+)
+def test_invalid_k_value(options, static):
+    """Tests if exception is thrown when passing
+    an invalid value for k when using a Random topology"""
+    with pytest.raises(ValueError):
+        GeneralOptimizerPSO(5, 2, options, Random(static=static))
+
+
+@pytest.mark.parametrize(
+    "options",
+    [
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": -1},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 6},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 0.5}
+    ],
+)
+def test_invalid_k_value(options):
+    """Tests if exception is thrown when passing
+    an invalid value for k when using a Random topology"""
+    with pytest.raises(ValueError):
+        GeneralOptimizerPSO(5, 2, options, Random())
+
+
+@pytest.mark.parametrize(
+    "topology",
+    [object(), int(), dict()]
+)
+def test_topology_type_exception(options, topology):
+    """Tests if exceptions are thrown when the topology has the wrong type"""
+    with pytest.raises(TypeError):
+        GeneralOptimizerPSO(5, 2, options, topology)
+
+
+@pytest.mark.parametrize(
+    "topology",
+    [Star(), Ring(), Pyramid(), Random()]
+)
+@pytest.mark.parametrize(
+    "bounds",
+    [
+        tuple(np.array([-5, -5])),
+        (np.array([-5, -5, -5]), np.array([5, 5])),
+        (np.array([-5, -5, -5]), np.array([5, 5, 5])),
+    ],
+)
+def test_bounds_size_exception(bounds, options, topology):
+    """Tests if exceptions are raised when bound sizes are wrong"""
+    with pytest.raises(IndexError):
+        GeneralOptimizerPSO(5, 2, options=options, topology=topology, bounds=bounds)
+
+
+@pytest.mark.parametrize(
+    "topology",
+    [Star(), Ring(), Pyramid(), Random()]
```

Comment:
Maybe we should use the `static` versions in this and the following tests as well.
It would also work if we create a variable:
```python
topologies = [Star(), Ring(), Ring(static=True), etc.]
```

And then use it in the decorator like so:
```python
@pytest.mark.parametrize("topology", topologies)
```

It would spare a lot of copy-pasting.

Reply:
Yes that’s correct

Reply:
I noticed that we already have a parametrized fixture for all the topologies so no need for the variable 👍